lib/excon/ssl_socket.rb

```diff
@@ -66,16 +66,22 @@ def initialize(data = {})
       end
 
       # maintain existing API
-      certificate_data = @data[:client_cert] || @data[:certificate_path]
-      private_key_data = @data[:client_key] || @data[:private_key_path]
-      private_key_pass = @data[:client_key_pass] || @data[:private_key_pass]
+      client_cert_data = if cert_data = @data[:client_cert_data]
+                           cert_data
+                         elsif path = (@data[:client_cert] || @data[:certificate_path])
+                           File.read path
+                         end
+
+      private_key_data = if key_data = @data[:private_key_data]
```

Comment:
Thanks for working with me on this and offering feedback :)!

I am unclear how my `if elsif` is different from your `this || (that || that)` snip above wrt precedence. Are you saying that you want it to read easier for people to see what we prefer? E.g. We prefer this (`client_cert_data`), or this (`client_cert`), but this last one (`certificate_path`) we don't really want cause it is the old API.

If so, would something like this be interesting (I am using extract method in order to clean up the body of `initialize`):
```diff
diff --git a/lib/excon/ssl_socket.rb b/lib/excon/ssl_socket.rb
index bbc8ac1..3df7664 100644
--- a/lib/excon/ssl_socket.rb
+++ b/lib/excon/ssl_socket.rb
@@ -65,34 +65,23 @@ module Excon
         ssl_context.verify_mode = OpenSSL::SSL::VERIFY_NONE
       end
 
-      # maintain existing API
-      client_cert_data = if cert_data = @data[:client_cert_data]
-                           cert_data
-                         elsif path = (@data[:client_cert] || @data[:certificate_path])
-                           File.read path
-                         end
-
-      private_key_data = if key_data = @data[:private_key_data]
-                           key_data
-                         elsif path = (@data[:client_key] || @data[:private_key_path])
-                           File.read path
-                         end
-
-      private_key_pass = @data[:client_key_pass] || @data[:private_key_pass]
-
-      if client_cert_data && private_key_data
+      client_cert_data = cert_data
+      client_key_data = key_data
+      client_key_pass = @data[:client_key_pass] || @data[:private_key_pass]
+
+      if client_cert_data && client_key_data
         ssl_context.cert = OpenSSL::X509::Certificate.new client_cert_data
         if OpenSSL::PKey.respond_to? :read
-          ssl_context.key = OpenSSL::PKey.read(private_key_data, private_key_pass)
+          ssl_context.key = OpenSSL::PKey.read(client_key_data, client_key_pass)
         else
-          ssl_context.key = OpenSSL::PKey::RSA.new(private_key_data, private_key_pass)
+          ssl_context.key = OpenSSL::PKey::RSA.new(client_key_data, client_key_pass)
         end
       elsif @data.key?(:certificate) && @data.key?(:private_key)
         ssl_context.cert = OpenSSL::X509::Certificate.new(@data[:certificate])
         if OpenSSL::PKey.respond_to? :read
-          ssl_context.key = OpenSSL::PKey.read(@data[:private_key], private_key_pass)
+          ssl_context.key = OpenSSL::PKey.read(@data[:private_key], client_key_pass)
         else
-          ssl_context.key = OpenSSL::PKey::RSA.new(@data[:private_key], private_key_pass)
+          ssl_context.key = OpenSSL::PKey::RSA.new(@data[:private_key], client_key_pass)
         end
       end
 
@@ -151,10 +140,33 @@ module Excon
 
     private
 
+    def cert_data
+      if cert_data = @data[:client_cert_data]
+        cert_data
+      elsif path = @data[:client_cert]
+        File.read path
+      elsif path = @data[:certificate_path]
+        warn ":certificate_path is no longer supported and will be deprecated. Please use :client_cert or :client_cert_data"
+        File.read path
+      end
+    end
+
     def connect
       # backwards compatability for things lacking nonblock
       @nonblock = HAVE_NONBLOCK && @nonblock
       super
     end
+
+    def key_data
+      if key_data = @data[:client_key_data]
+        key_data
+      elsif path = @data[:client_key]
+        File.read path
+      elsif path = @data[:private_key_path]
+        warn ":private_key_path is no longer supported and will be deprecated. Please use :client_key or :client_key_data"
+        File.read path
+      end
+    end
+
   end
 end
```
We could also reduce local variables by setting the `*_data` methods to cached instance variables if you would like.
```diff
diff --git a/lib/excon/ssl_socket.rb b/lib/excon/ssl_socket.rb
index bbc8ac1..ea1c301 100644
--- a/lib/excon/ssl_socket.rb
+++ b/lib/excon/ssl_socket.rb
@@ -65,34 +65,21 @@ module Excon
         ssl_context.verify_mode = OpenSSL::SSL::VERIFY_NONE
       end
 
-      # maintain existing API
-      client_cert_data = if cert_data = @data[:client_cert_data]
-                           cert_data
-                         elsif path = (@data[:client_cert] || @data[:certificate_path])
-                           File.read path
-                         end
-
-      private_key_data = if key_data = @data[:private_key_data]
-                           key_data
-                         elsif path = (@data[:client_key] || @data[:private_key_path])
-                           File.read path
-                         end
-
-      private_key_pass = @data[:client_key_pass] || @data[:private_key_pass]
-
-      if client_cert_data && private_key_data
+      client_key_pass = @data[:client_key_pass] || @data[:private_key_pass]
+
+      if client_cert_data && client_key_data
         ssl_context.cert = OpenSSL::X509::Certificate.new client_cert_data
         if OpenSSL::PKey.respond_to? :read
-          ssl_context.key = OpenSSL::PKey.read(private_key_data, private_key_pass)
+          ssl_context.key = OpenSSL::PKey.read(client_key_data, client_key_pass)
         else
-          ssl_context.key = OpenSSL::PKey::RSA.new(private_key_data, private_key_pass)
+          ssl_context.key = OpenSSL::PKey::RSA.new(client_key_data, client_key_pass)
         end
       elsif @data.key?(:certificate) && @data.key?(:private_key)
         ssl_context.cert = OpenSSL::X509::Certificate.new(@data[:certificate])
         if OpenSSL::PKey.respond_to? :read
-          ssl_context.key = OpenSSL::PKey.read(@data[:private_key], private_key_pass)
+          ssl_context.key = OpenSSL::PKey.read(@data[:private_key], client_key_pass)
         else
-          ssl_context.key = OpenSSL::PKey::RSA.new(@data[:private_key], private_key_pass)
+          ssl_context.key = OpenSSL::PKey::RSA.new(@data[:private_key], client_key_pass)
         end
       end
 
@@ -151,10 +138,33 @@ module Excon
 
     private
 
+    def client_cert_data
+      @client_cert_data ||= if ccd = @data[:client_cert_data]
+                              ccd
+                            elsif path = @data[:client_cert]
+                              File.read path
+                            elsif path = @data[:certificate_path]
+                              warn ":certificate_path is no longer supported and will be deprecated. Please use :client_cert or :client_cert_data"
+                              File.read path
+                            end
+    end
+
     def connect
       # backwards compatability for things lacking nonblock
       @nonblock = HAVE_NONBLOCK && @nonblock
       super
     end
+
+    def client_key_data
+      @client_key_data ||= if ckd = @data[:client_key_data]
+                             ckd
+                           elsif path = @data[:client_key]
+                             File.read path
+                           elsif path = @data[:private_key_path]
+                             warn ":private_key_path is no longer supported and will be deprecated. Please use :client_key or :client_key_data"
+                             File.read path
+                           end
+    end
+
   end
 end
```
I didn't do that though because I am not sure I want people to be able to easily print their `*.pem`s if they are `puts`ing the connection or something. Not sure if that matters or not.

Maybe we could even extract method on the key pass too:
```diff
diff --git a/lib/excon/ssl_socket.rb b/lib/excon/ssl_socket.rb
index bbc8ac1..ea78bd2 100644
--- a/lib/excon/ssl_socket.rb
+++ b/lib/excon/ssl_socket.rb
@@ -78,21 +78,19 @@ module Excon
                            File.read path
                          end
 
-      private_key_pass = @data[:client_key_pass] || @data[:private_key_pass]
-
       if client_cert_data && private_key_data
         ssl_context.cert = OpenSSL::X509::Certificate.new client_cert_data
         if OpenSSL::PKey.respond_to? :read
-          ssl_context.key = OpenSSL::PKey.read(private_key_data, private_key_pass)
+          ssl_context.key = OpenSSL::PKey.read(private_key_data, client_key_pass)
         else
-          ssl_context.key = OpenSSL::PKey::RSA.new(private_key_data, private_key_pass)
+          ssl_context.key = OpenSSL::PKey::RSA.new(private_key_data, client_key_pass)
         end
       elsif @data.key?(:certificate) && @data.key?(:private_key)
         ssl_context.cert = OpenSSL::X509::Certificate.new(@data[:certificate])
         if OpenSSL::PKey.respond_to? :read
-          ssl_context.key = OpenSSL::PKey.read(@data[:private_key], private_key_pass)
+          ssl_context.key = OpenSSL::PKey.read(@data[:private_key], client_key_pass)
         else
-          ssl_context.key = OpenSSL::PKey::RSA.new(@data[:private_key], private_key_pass)
+          ssl_context.key = OpenSSL::PKey::RSA.new(@data[:private_key], client_key_pass)
         end
       end
 
@@ -151,6 +149,10 @@ module Excon
 
     private
 
+    def client_key_pass
+      @data[:client_key_pass] || @data[:private_key_pass]
+    end
+
     def connect
       # backwards compatability for things lacking nonblock
       @nonblock = HAVE_NONBLOCK && @nonblock
```